File: sling/nlp/parser/trainer/transition-generator.cc
// Copyright 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "sling/nlp/parser/trainer/transition-generator.h"

#include <algorithm>
#include <unordered_set>

#include "sling/base/macros.h"
#include "sling/string/strcat.h"

namespace sling {
namespace nlp {

std::vector<string> TransitionSequence::AsStrings(Store *store) const {
  std::vector<string> debug;
  for (const ParserAction &action : actions()) {
    debug.emplace_back(action.ToString(store));
  }

  return debug;
}

int TransitionSequence::FirstAction(int token) const {
  // The desired action is immediately after the SHIFT action for the
  // previous token. Since there is exactly one SHIFT per token, we simply
  // count (token - 1) SHIFTs.
  DCHECK_GT(actions_.size(), 0);
  if (token == 0) return 0;

  int skips = 0;
  for (int i = 0; i < actions_.size(); ++i) {
    if (actions_[i].type == ParserAction::SHIFT) ++skips;
    if (skips == token) return i + 1;
  }

  return actions_.size();  // should not reach here
}

void TransitionGenerator::AttentionIndex::Add(Handle handle) {
  const auto &it = frame_to_index_.find(handle);
  if (it == frame_to_index_.end()) {
    index_to_frame_.push_back(handle);
    frame_to_index_[handle] = index_to_frame_.size() - 1;
  }
  MakeCenter(handle);
}

void TransitionGenerator::AttentionIndex::MakeCenter(Handle handle) {
  int current_index = Index(handle);
  if (current_index == 0) return;  // already the center

  for (int i = current_index; i >= 0; --i) {
    index_to_frame_[i] = (i == 0) ? handle : index_to_frame_[i - 1];
    frame_to_index_[index_to_frame_[i]] = i;
  }
}

void TransitionGenerator::AttentionIndex::Update(const Action &action) {
  switch (action.type) {
    case ParserAction::EVOKE: {
      Add(action.frame->handle);
      break;
    }
    case ParserAction::REFER: {
      MakeCenter(action.frame->handle);
      break;
    }
    case ParserAction::EMBED: {
      Add(action.frame->handle);
      break;
    }
    case ParserAction::ELABORATE: {
      Add(action.frame->handle);
      break;
    }
    case ParserAction::CONNECT: {
      MakeCenter(action.frame->handle);
      break;
    }
    case ParserAction::ASSIGN: {
      MakeCenter(action.frame->handle);
      break;
    }
    case ParserAction::SHIFT:
    case ParserAction::STOP:
      break;
    default:
      LOG(FATAL) << "Unknown action";
      break;
  }
}

void TransitionGenerator::InitInfo(const Frame &frame,
                                   HandleMap<FrameInfo *> *frame_info,
                                   HandleSet *initialized) const {
  Handle handle = frame.handle();
  if (initialized->find(handle) != initialized->end()) return;
  initialized->insert(handle);

  FrameInfo *&info = (*frame_info)[handle];
  if (info == nullptr) info = new FrameInfo();
  info->handle = handle;
  std::vector<Frame> pending;  // frames whose FrameInfo needs to be initialized
  std::unordered_set<uint64> seen;  // processed slots
  for (const Slot &slot : frame) {
    if (slot.name.IsId() || slot.name == n_name_.handle()) continue;

    // Ignore local slot names. We are only interested in global roles.
    if (slot.name.IsLocalRef()) continue;
    if (slot.name.IsIsA() && !slot.value.IsGlobalRef()) continue;

    // Ignore duplicates.
    uint64 key = slot.name.raw();
    key = (key << 32) | slot.value.raw();
    if (seen.find(key) != seen.end()) continue;
    seen.insert(key);

    if (slot.name.IsIsA() && info->first_type.IsNil()) {
      info->first_type = slot.value;
    } else {
      info->edges.emplace_back(false /* incoming */, slot.name, slot.value);

      // Check for self-loops.
      if (frame.handle() == slot.value) {
        info->edges.back().inverse = info->edges.size() - 1;
        continue;
      }

      // Add incoming edge to the linked frame's FrameInfo only if the linked
      // frame is local. We don't want to output EVOKE/EMBED for global frames.
      Object target(frame.store(), slot.value);
      if (target.IsFrame() && target.IsLocal()) {
        FrameInfo *&neighbor_info = (*frame_info)[slot.value];
        if (neighbor_info == nullptr) neighbor_info = new FrameInfo();

        auto &edges = neighbor_info->edges;
        edges.emplace_back(true /* incoming */, slot.name, handle);
        edges.back().inverse = info->edges.size() - 1;
        info->edges.back().inverse = edges.size() - 1;
        pending.emplace_back(target.AsFrame());
      }
    }
  }

  // Set a default type for the frame.
  if (info->first_type.IsNil()) info->first_type = n_thing_.handle();

  // Recursively process all local frames that this frame points to.
  for (const Frame &f : pending) InitInfo(f, frame_info, initialized);
}

ParserAction TransitionGenerator::AttentionIndex::Translate(
    const Document &document,
    const Action &action) {
  ParserAction output;
  output.type = action.type;
  switch (action.type) {
    case ParserAction::EVOKE: {
      CHECK_NE(action.length, 0);
      output.length = action.length;
      output.label = action.frame->first_type;
      break;
    }
    case ParserAction::REFER: {
      CHECK_NE(action.length, 0);
      output.length = action.length;
      output.target = Index(action.frame->handle);
      SetMaxIndex(output.target);
      break;
    }
    case ParserAction::EMBED: {
      output.label = action.frame->first_type;
      output.role = action.role;
      output.target = Index(action.other_frame->handle);
      SetMaxIndex(output.target);
      break;
    }
    case ParserAction::ELABORATE: {
      output.label = action.frame->first_type;
      output.role = action.role;
      output.source = Index(action.other_frame->handle);
      SetMaxIndex(output.source);
      break;
    }
    case ParserAction::CONNECT: {
      output.source = Index(action.frame->handle);
      output.target = Index(action.other_frame->handle);
      output.role = action.role;
      SetMaxIndex(output.source);
      SetMaxIndex(output.target);
      break;
    }
    case ParserAction::ASSIGN: {
      output.source = Index(action.frame->handle);
      output.role = action.role;
      output.label = action.value;
      SetMaxIndex(output.source);
      break;
    }
    case ParserAction::SHIFT:
    case ParserAction::STOP:
    default:
      break;
  }

  return output;
}

void TransitionGenerator::Generate(
    const Document &document,
    TransitionSequence *sequence,
    TransitionGenerator::Report *report) const {
  Generate(document, 0, document.num_tokens(), sequence, report);
}

void TransitionGenerator::Generate(
    const Document &document,
    int begin,
    int end,
    TransitionSequence *sequence,
    TransitionGenerator::Report *report) const {
  Store *store = document.store();

  // Sort the spans by start token, and then in decreasing order of length.
  // This ordering is used to EVOKE/REFER the corresponding evoked frames.
  std::vector<Span *> spans;
  for (int i = 0; i < document.num_spans(); ++i) {
    Span *span = document.span(i);
    if (span->begin() >= begin && span->end() <= end) spans.emplace_back(span);
  }
  std::sort(spans.begin(), spans.end(), [](Span *s, Span *t) {
    return (s->begin() < t->begin()) ||
        ((s->begin() == t->begin()) && (s->end() > t->end()));
  });

  // Frame handle -> FrameInfo.
  HandleMap<FrameInfo *> frame_info;

  // Initialize FrameInfo of all evoked frames.
  HandleSet initialized;
  std::vector<Span *> spans_with_frames;
  for (Span *span : spans) {
    for (const Slot &slot : span->mention()) {
      if (slot.name == n_evokes_.handle()) {
        if (spans_with_frames.empty() || spans_with_frames.back() != span) {
          spans_with_frames.emplace_back(span);
        }
        InitInfo(Frame(store, slot.value), &frame_info, &initialized);
      }
    }
  }

  // Initialize FrameInfo of all thematic frames.
  for (Handle theme : document.themes()) {
    Object object(store, theme);
    if (object.IsFrame()) {
      InitInfo(object.AsFrame(), &frame_info, &initialized);
    }
  }

  // Insert SHIFT actions between consecutive EVOKE/REFER actions. Other
  // actions triggered by EVOKE will be added to the list later.
  int previous_start = begin;
  std::vector<Action> stack;
  HandleSet evoked;
  for (Span *span : spans_with_frames) {
    // Issue SHIFT actions from previous span till this span. This includes
    // skipping all tokens for the previous span.
    for (int i = previous_start; i < span->begin(); ++i) {
      stack.emplace_back(ParserAction::SHIFT);
    }
    previous_start = span->begin();

    // Issue an EVOKE/REFER action for every evoked frame.
    for (const Slot &slot : span->mention()) {
      if (slot.name == n_evokes_.handle()) {
        FrameInfo *f = frame_info[slot.value];
        if (evoked.count(slot.value) == 0) {
          stack.emplace_back(ParserAction::EVOKE, f, span->length());
          evoked.insert(slot.value);
        } else {
          stack.emplace_back(ParserAction::REFER, f, span->length());
        }
      }
    }
  }

  // Add SHIFTs for tokens after the last frame, and the final STOP action.
  for (int i = previous_start; i < end; ++i) {
    stack.emplace_back(ParserAction::SHIFT);
  }
  stack.emplace_back(ParserAction::STOP);

  // Reverse 'stack' and treat it as a stack.
  // This is for simplicity -- e.g. we can pop an EVOKE action from the stack,
  // and push triggered actions like CONNECT, ASSIGN etc. onto the stack.
  // Note that reversing + treating as stack continues to preserve span-ordering
  // of evoked frames.
  std::reverse(stack.begin(), stack.end());

  AttentionIndex attention;
  while (!stack.empty()) {
    Action action = stack.back();
    stack.pop_back();

    // Translate the action using real attention indices and output it.
    sequence->Add(attention.Translate(document, action));

    // Update the attention index by performing the action.
    attention.Update(action);

    // New frame creation can trigger more actions. Add them to the stack.
    if (action.type == ParserAction::EVOKE ||
        action.type == ParserAction::EMBED ||
        action.type == ParserAction::ELABORATE) {
      action.frame->output = true;

      // Process CONNECT actions.
      for (auto &edge : action.frame->edges) {
        if (edge.used) continue;
        const auto &it = frame_info.find(edge.neighbor);
        if (it == frame_info.end()) continue;

        FrameInfo *neighbor_info = it->second;
        if (neighbor_info->output) {
          // Neighbor has already been output earlier. Output a CONNECT action.
          stack.emplace_back(ParserAction::CONNECT);
          Action &latest = stack.back();
          latest.frame = edge.incoming ? neighbor_info : action.frame,
          latest.role = edge.role;
          latest.other_frame = edge.incoming ? action.frame : neighbor_info;
          edge.used = true;
          CHECK_NE(edge.inverse, -1);
          neighbor_info->edges[edge.inverse].used = true;
        }
      }

      // Process EMBED actions.
      for (auto &edge : action.frame->edges) {
        if (edge.used) continue;
        const auto &it = frame_info.find(edge.neighbor);
        if (it == frame_info.end()) continue;

        // EMBED a frame if it links to this frame, hasn't been output earlier,
        // and if it is not evoked by any span.
        FrameInfo *neighbor_info = it->second;
        if (!neighbor_info->output && edge.incoming) {
          const auto &range = document.EvokingSpans(neighbor_info->handle);
          if (range.begin() == range.end()) {
            stack.emplace_back(ParserAction::EMBED);
            Action &latest = stack.back();
            latest.frame = neighbor_info;
            latest.role = edge.role;
            latest.other_frame = action.frame;
            edge.used = true;
            neighbor_info->edges[edge.inverse].used = true;
          }
        }
      }

      // Process ELABORATE actions.
      for (auto &edge : action.frame->edges) {
        if (edge.used) continue;
        const auto &it = frame_info.find(edge.neighbor);
        if (it == frame_info.end()) continue;

        // ELABORATE this frame by creating new frames and linking to them.
        // Links must be TO the new frames, and the new frames must not be
        // evoked by spans.
        FrameInfo *neighbor_info = it->second;
        const auto &range = document.EvokingSpans(neighbor_info->handle);
        if (range.begin() == range.end() &&
            !neighbor_info->output &&
            !edge.incoming) {
          stack.emplace_back(ParserAction::ELABORATE, neighbor_info);
          Action &latest = stack.back();
          latest.role = edge.role;
          latest.other_frame = action.frame;
          edge.used = true;
          neighbor_info->edges[edge.inverse].used = true;
        }
      }

      // Process ASSIGN actions.
      // Note: Since we push ASSIGN actions on the stack last, they are the
      // first one to be popped out. This ensures that ASSIGN's index argument
      // will always be zero.
      for (auto &edge : action.frame->edges) {
        if (edge.used) continue;

        Object neighbor(store, edge.neighbor);
        if (!neighbor.IsLocal()) {
          // Treat global values as constants & output ASSIGN actions for them.
          DCHECK(!edge.incoming);
          stack.emplace_back(ParserAction::ASSIGN, action.frame);
          Action &latest = stack.back();
          latest.role = edge.role;
          latest.value = edge.neighbor;
          edge.used = true;
        }
      }
    }
  }

  // Output the report.
  if (report != nullptr) {
    report->max_attention_index = attention.MaxIndex();
    for (const auto &kv : frame_info) {
      Handle frame_handle = kv.first;
      const FrameInfo &info = *kv.second;
      bool frame_output = info.output;
      if (!frame_output) {
        // Frame was never pushed into the attention buffer.
        Frame frame(store, frame_handle);
        report->frames_not_output++;
        StrAppend(&report->not_output_debug, "Frame:", frame.Id(), "\n");
      }

      // Check for any unprocessed edges.
      for (const auto &edge : info.edges) {
        if (!edge.incoming && !edge.used) {
          report->edges_not_output++;
          if (frame_output) {
            StrAppend(&report->not_output_debug,
                      Object(store, frame_handle).DebugString(), " ->(",
                      Object(store, edge.role).DebugString(), ")-> ",
                      Object(store, edge.neighbor).DebugString(), "\n");
          }
        }
      }
    }
  }

  // Cleanup.
  for (auto &kv : frame_info) delete kv.second;
}

}  // namespace nlp
}  // namespace sling

File: sling/nlp/parser/trainer/transition_generator.py
# Copyright 2017 Google Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http:#www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import sling
from action import Action

# Outputs a list of transitions that represent a given document's frame graph.
class TransitionGenerator:
  # Bookkeeping for one frame.
  class FrameInfo:
    def __init__(self, handle):
      self.handle = handle
      self.type = None
      self.edges = []
      self.from_mention = False

      # Whether this frame has been evoked.
      self.output = False


  # Labeled edge between two frames. Each edge is used to issue a CONNECT,
  # EMBED, ELABORATE, or ASSIGN action.
  class Edge:
    def __init__(self, incoming=None, role=None, value=None):
      self.incoming = incoming
      self.role = role
      self.neighbor = value
      self.inverse = None
      self.used = False


  # Simple action that will be translated into an Action object later on.
  class SimpleAction:
    def __init__(self, t=None):
      self.action = Action(t)
      self.info = None
      self.other_info = None


  def __init__(self, commons):
    self.commons = commons
    self._id = commons["id"]
    self._isa = commons["isa"]


  # Returns whether 'handle' refers to another frame.
  def is_ref(self, handle):
    return type(handle) == sling.Frame


  # Creates a FrameInfo object for 'frame' and recursively for all frames
  # pointed to by it.
  def _init_info(self, frame, frame_info, initialized):
    if frame in initialized:
      return
    initialized[frame] = True

    info = frame_info.get(frame, None)
    if info is None:
      info = TransitionGenerator.FrameInfo(frame)
      frame_info[frame] = info

    pending = []
    for role, value in frame:
      if not self.is_ref(value) or role == self._id: continue
      if role == self._isa and value.islocal(): continue

      if role == self._isa and info.type is None:
        info.type = value
      else:
        edge = TransitionGenerator.Edge(incoming=False, role=role, value=value)
        info.edges.append(edge)
        if value == frame:
          edge.inverse = edge
          continue

        if value.islocal():
          nb_info = frame_info.get(value, None)
          if nb_info is None:
            nb_info = TransitionGenerator.FrameInfo(value)
            frame_info[value] = nb_info
          nb_edge = TransitionGenerator.Edge(
              incoming=True, role=role, value=frame)
          nb_info.edges.append(nb_edge)
          nb_edge.inverse = edge
          edge.inverse = nb_edge
          pending.append(value)

    # Initialize bookkeeping for all frames pointed to by this frame.
    for p in pending:
      self._init_info(p, frame_info, initialized)


  # Translates 'simple' action to an Action using indices from 'attention'.
  def _translate(self, attention, simple):
    action = Action(t=simple.action.type)
    if simple.action.length is not None:
      action.length = simple.action.length
    if simple.action.role is not None:
      action.role = simple.action.role

    if action.type == Action.EVOKE:
       action.label = simple.info.type
    elif action.type == Action.REFER:
      action.target = attention.index(simple.info.handle)
    elif action.type == Action.EMBED:
      action.label = simple.info.type
      action.target = attention.index(simple.other_info.handle)
    elif action.type == Action.ELABORATE:
      action.label = simple.info.type
      action.source = attention.index(simple.other_info.handle)
    elif action.type == Action.CONNECT:
      action.source = attention.index(simple.info.handle)
      action.target = attention.index(simple.other_info.handle)
    elif action.type == Action.ASSIGN:
      action.source = attention.index(simple.info.handle)
      action.label = simple.action.label

    return action


  # Updates frame indices in 'attention' as a result of the action 'simple'.
  def _update(self, attention, simple):
    t = simple.action.type
    if t in [Action.EVOKE, Action.EMBED, Action.ELABORATE]:
      # Insert a new frame at the center of attention.
      attention.insert(0, simple.info.handle)
    elif t in [Action.REFER, Action.ASSIGN, Action.CONNECT]:
      # Promote an existing frame to the center of attention.
      attention.remove(simple.info.handle)
      attention.insert(0, simple.info.handle)


  # Builds and returns a simple action of type 'type'.
  def _simple_action(self, type=None):
    return TransitionGenerator.SimpleAction(type)


  # Stores mentions starting or ending or both at a given token.
  class TokenToMentions:
    def __init__(self):
      self.starting = []
      self.ending = []
      self.singletons = []

    # Record 'mention' at starting at this token.
    def start(self, mention):
      if len(self.starting) > 0:
        # Check that the mention respects nesting.
        assert self.starting[-1].end >= mention.end
      self.starting.append(mention)

    # Record 'mention' as ending at this token.
    def end(self, mention):
      if len(self.ending) > 0:
        # Check that the mention respects nesting.
        assert self.ending[0].begin <= mention.begin
      self.ending.insert(0, mention)  # most-nested is at the front

    # Record 'mention' as starting and ending at this token.
    def singleton(self, mention):
      self.singletons.append(mention)

    # Returns if there are no mentions starting/ending at this token.
    def empty(self):
      return len(self.starting) + len(self.ending) + len(self.singletons) == 0

    # Returns a string representation of the object.
    def __repr__(self):
      return "Starting:" + str(self.starting) + ", Ending:" + \
        str(self.ending) + ", Singletons:" + str(self.singletons)


  # Generates transition sequence for 'document' which should be an instance of
  # AnnotatedDocument.
  def generate(self, document):
    frame_info = {}    # frame -> whether it is evoked from a span
    initialized = {}   # frame -> whether the frame's book-keeping is done

    # Initialize book-keeping for all evoked frames.
    for m in document.mentions:
      for evoked in m.evokes():
        self._init_info(evoked, frame_info, initialized)
        frame_info[evoked].from_mention = True

    # Initialize book-keeping for all thematic frames.
    for theme in document.themes:
      self._init_info(theme, frame_info, initialized)

    # Record start/end boundaries of all mentions.
    token_to_mentions = []
    for _ in range(len(document.tokens)):
      token_to_mentions.append(TransitionGenerator.TokenToMentions())

    for m in document.mentions:
      if m.length == 1:
        token_to_mentions[m.begin].singleton(m)
      else:
        token_to_mentions[m.begin].start(m)
        token_to_mentions[m.end - 1].end(m)

    # Single token mentions are handled via EVOKE(length=1), and others
    # are handled via MARK at the beginning token and EVOKE(length=None)
    # at the end token.

    simple_actions = []
    marked = {}   # frames for which we have output a MARK
    evoked = {}   # frames for which we have output an EVOKE
    for index in range(len(document.tokens)):
      t2m = token_to_mentions[index]

      # First evoke/refer the singleton mentions.
      for singleton in t2m.singletons:
        for frame in singleton.evokes():
          # If the frame is already evoked, refer to it.
          if frame in marked:
            assert frame in evoked, "Referring to marked but not evoked frame"
          if frame in evoked:
            refer = self._simple_action(Action.REFER)
            refer.info = frame_info[frame]
            refer.action.length = singleton.length  # should be 1
            simple_actions.append(refer)
            continue

          # Otherwise evoke a new frame.
          evoke = self._simple_action(Action.EVOKE)
          evoke.action.length = singleton.length  # should be 1
          evoke.info = frame_info[frame]
          evoke.action.label = evoke.info.type
          simple_actions.append(evoke)
          marked[frame] = True
          evoked[frame] = True

      # Output EVOKE for any frames whose spans end here.
      for mention in t2m.ending:
        assert mention.length > 1, mention.length  # singletons already handled
        for frame in mention.evokes():
          assert frame in marked   # frame should be already MARKed
          if frame in evoked:
            # Already handled via REFER at mention.begin.
            continue

          evoke = self._simple_action(Action.EVOKE)
          evoke.info = frame_info[frame]
          evoke.action.label = evoke.info.type
          simple_actions.append(evoke)
          evoked[frame] = True

      # Output MARK for any frames whose spans begin here.
      for mention in t2m.starting:
        assert mention.length > 1, mention.length  # singletons already handled
        for frame in mention.evokes():
          # Check if this is a fresh frame or a refer.
          if frame in marked:
            assert frame in evoked, "Referring to marked but not evoked frame"
          if frame in evoked:
            refer = self._simple_action(Action.REFER)
            refer.info = frame_info[frame]
            refer.action.length = mention.length
            simple_actions.append(refer)
            continue

          mark = self._simple_action(Action.MARK)
          mark.info = frame_info[frame]
          simple_actions.append(mark)
          marked[frame] = True

      # Move to the next token.
      simple_actions.append(self._simple_action(Action.SHIFT))
    simple_actions.append(self._simple_action(Action.STOP))

    # Recursively output more actions (e.g. CONNECT, EMBED, ELABORATE, ASSIGN)
    # from the current set of EVOKE/REFER actions. Then translate each
    # action using final attention indices. This is done in reverse order
    # for convenience.
    simple_actions.reverse()
    actions = []
    attention = []
    while len(simple_actions) > 0:
      simple_action = simple_actions.pop()
      actions.append(self._translate(attention, simple_action))
      self._update(attention, simple_action)
      t = simple_action.action.type
      if t == Action.EVOKE or t == Action.EMBED or t == Action.ELABORATE:
        simple_action.info.output = True

        # CONNECT actions triggered by the newly output frame.
        for e in simple_action.info.edges:
          if e.used: continue

          nb = frame_info.get(e.neighbor, None)
          if nb is not None and nb.output:
            connect = self._simple_action(Action.CONNECT)
            connect.action.role = e.role
            connect.info = nb if e.incoming else simple_action.info
            connect.other_info = simple_action.info if e.incoming else nb
            simple_actions.append(connect)
            e.used = True
            e.inverse.used = True

        # EMBED actions triggered by the newly output frame.
        for e in simple_action.info.edges:
          if e.used or not e.incoming: continue

          nb = frame_info.get(e.neighbor, None)
          if nb is not None and not nb.output and not nb.from_mention:
            embed = self._simple_action(Action.EMBED)
            embed.action.role = e.role
            embed.info = nb
            embed.other_info = simple_action.info
            simple_actions.append(embed)
            e.used = True
            e.inverse.used = True

        # ELABORATE actions triggered by the newly output frame.
        for e in simple_action.info.edges:
          if e.used or e.incoming: continue

          nb = frame_info.get(e.neighbor, None)
          if nb is not None and not nb.output and not nb.from_mention:
            elaborate = self._simple_action(Action.ELABORATE)
            elaborate.action.role = e.role
            elaborate.info = nb
            elaborate.other_info = simple_action.info
            simple_actions.append(elaborate)
            e.used = True
            e.inverse.used = True

        # ASSIGN actions triggered by the newly output frame.
        for e in simple_action.info.edges:
          if not e.used and not e.neighbor.islocal() and not e.incoming:
            assign = self._simple_action(Action.ASSIGN)
            assign.info = simple_action.info
            assign.action.role = e.role
            assign.action.label = e.neighbor
            simple_actions.append(assign)
            e.used = True

    return actions

